snorkel/contrib/deps/labeling/model/label_model.py

```diff
@@ -0,0 +1,180 @@
+
+        prob = cp.Problem(objective, constraints)
+        prob.solve(verbose=False)
+        U, s, V = np.linalg.svd(L_cvx.value)
+        Z = np.sqrt(s[: self.cardinality]) * U[:, : self.cardinality]
+        O = self.O.numpy()
+        I_k = np.eye(self.cardinality)
+        return O @ Z @ np.linalg.inv(I_k + Z.T @ O @ Z) @ Z.T @ O
+
+    def _fit_loss(self, loss_fn):
+        # Restore model if necessary
+        start_iteration = 0
+
+        # Train the model
+        metrics_hist = {}  # The most recently seen value for all metrics
+        for epoch in range(start_iteration, self.train_config.n_epochs):
+            self.running_loss = 0.0
+            self.running_examples = 0
+
+            # Zero the parameter gradients
+            self.optimizer.zero_grad()
+
+            # Forward pass to calculate the average loss per example
+            loss = self._loss_mu(l2=self.train_config.l2)
+            if torch.isnan(loss):
+                msg = "Loss is NaN. Consider reducing learning rate."
+                raise Exception(msg)
+
+            # Backward pass to calculate gradients
+            # Loss is an average loss per example
+            loss.backward()
+
+            # Perform optimizer step
+            self.optimizer.step()
+
+            # Calculate metrics, log, and checkpoint as necessary
+            metrics_dict = self._execute_logging(loss)
+            metrics_hist.update(metrics_dict)
+
+            # Update learning rate
+            self._update_lr_scheduler(epoch)
+
+    def fit_with_deps(
+        self,
+        L_train: np.ndarray,
+        Y_dev: Optional[np.ndarray] = None,
+        deps: Optional[List[Tuple[int, int]]] = None,
+        class_balance: Optional[List[float]] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Train label model using dependencies (if given).
+
+        Train label model to estimate mu, the parameters used to combine LFs.
+
+        Parameters
+        ----------
+        L_train
+            An [n,m] matrix with values in {-1,0,1,...,k-1}
+        Y_dev
+            Gold labels for dev set for estimating class_balance, by default None
+        deps
+            Optional list of pairs of correlated LF indices.
+        class_balance
+            Each class's percentage of the population, by default None
+        **kwargs
+            Arguments for changing train config defaults
+
+        Raises
+        ------
+        Exception
+            If loss in NaN
+
+        Examples
+        --------
+        >>> L = np.array([[0, 0, -1], [-1, 0, 1], [1, -1, 0]])
+        >>> Y_dev = [0, 1, 0]
+        >>> label_model = DependencyAwareLabelModel(verbose=False)
+        >>> label_model.fit_with_deps(L, deps=[(0, 2)])  # doctest: +SKIP
+        >>> label_model.fit_with_deps(L, deps=[(0, 2)], Y_dev=Y_dev)  # doctest: +SKIP
+        >>> label_model.fit_with_deps(L, deps=[(0, 2)], class_balance=[0.7, 0.3])  # doctest: +SKIP
+        """
+        # Set random seed
```

Comment:
if `deps` passed in is empty, should it default to calling the original LabelModel fit function?

Reply:
We want this to eventually replace LabelModel right? So it'd be good if the current code can just be copy pasted in place of LabelModel.fit. Adding a call to LabelModel.fit will make it harder.